File: Codeforces/Practice/Power_of_three.py
def is_power_of_three(n: str) -> int:
    # Edge cases
    if not n or n == '0':
        return -1
    if n == '1':
        return 0
        
    # Remove leading zeros
    n = n.lstrip('0')
    if not n:
        return -1
        
    # Quick pattern checks
    if n[-1] not in '1379':
        return -1
    
    # Check divisibility by 3 using digital root
    if sum(int(d) for d in n) % 3 != 0:
        return -1
    
    # For very large numbers, use logarithmic method first
    from math import log10
    
    # Calculate approximate exponent using logarithms
    length = len(n)
    first_digit = int(n[0])
    log10_n = (length - 1) + log10(first_digit)
    log10_3 = log10(3)
    approximate_exp = log10_n / log10_3
    
    # If approximate_exp is far from an integer, not a power of 3
    if abs(approximate_exp - round(approximate_exp)) > 0.0001:
        return -1
    
    # For smaller numbers, verify directly
    if length < 19:  # Changed from 100 to handle numbers up to 10^18 safely
        num = int(n)
        count = 0
        while num > 1:
            if num % 3 != 0:
                return -1
            num //= 3
            count += 1
        return count
    
    # For larger numbers, use string division
    def divide_by_three(s):
        result = []
        carry = 0
        for digit in s:
            current = carry * 10 + int(digit)
            result.append(str(current // 3))
            carry = current % 3
        while len(result) > 1 and result[0] == '0':
            result.pop(0)
        return ''.join(result)
    
    # Verify by division
    current = n
    count = 0
    max_iterations = int(approximate_exp) + 2  # Add buffer for precision
    
    while current != '1':
        if count > max_iterations:
            return -1
        if not all(d in '0123456789' for d in current):
            return -1
        old_len = len(current)
        current = divide_by_three(current)
        if current == '0' or (count > 0 and current == n):
            return -1
        count += 1
        
    return count

# Fast I/O
if __name__ == "__main__":
    # Test cases
    assert is_power_of_three("9") == 2    # 3^2 = 9
    assert is_power_of_three("27") == 3   # 3^3 = 27
    assert is_power_of_three("45") == -1  # Not a power of 3
    assert is_power_of_three("243") == 5  # 3^5 = 243
    assert is_power_of_three("1") == 0    # 3^0 = 1
    print(is_power_of_three(input().strip()))

File: Codeforces/Practice/Power_of_three.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

// Quick check for invalid last digits of power of 3
bool hasValidLastDigit(const string &N)
{
    if (N.empty())
        return false;
    char last = N.back();
    return (last == '1' || last == '3' || last == '7' || last == '9');
}

// Optimized divisibility by 3 check using digital root
bool isDivisibleByThree(const string &N)
{
    // Digital root for divisibility by 3
    int sum = 0;
    for (char digit : N)
    {
        sum += digit - '0';
        // Keep sum small to handle very large numbers
        if (sum >= 30)
            sum %= 3;
    }
    return sum % 3 == 0;
}

string divideByThree(const string &N)
{
    string result;
    result.reserve(N.length()); // Pre-allocate memory

    int carry = 0;
    bool leading = true;

    // Process 18 digits at a time to avoid overflow
    // This is much faster for very large numbers
    for (size_t i = 0; i < N.length(); i++)
    {
        int current = carry * 10 + (N[i] - '0');
        int quotient = current / 3;
        if (!leading || quotient != 0)
        {
            result += (quotient + '0');
            leading = false;
        }
        carry = current % 3;
    }

    return result.empty() ? "0" : result;
}

int findPowerOfThree(const string &N)
{
    // Edge cases
    if (N.empty() || N == "0")
        return -1;
    if (N == "1")
        return 0;

    // Quick validation checks
    if (!hasValidLastDigit(N))
        return -1;
    if (!isDivisibleByThree(N))
        return -1;

    string current = N;
    int count = 0;
    const int MAX_ITERATIONS = 39; // log_3(10^18) ≈ 38.4

    while (current != "1")
    {
        // Additional optimization: check size reduction
        size_t oldSize = current.length();
        current = divideByThree(current);
        size_t newSize = current.length();

        // Each division by 3 should reduce length by roughly log_10(3) ≈ 0.477
        // If reduction is too small, number is not a power of 3
        if (oldSize > 20 && (oldSize - newSize) < (oldSize / 3))
        {
            return -1;
        }

        count++;
        // If we've done too many iterations, it's not a power of 3
        if (count > MAX_ITERATIONS)
            return -1;
    }

    return count;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    string N;
    cin >> N;
    cout << findPowerOfThree(N) << endl;

    return 0;
}